subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/logging/AbstractTestLogger.java

```diff
@@ -83,9 +83,10 @@ private String getEventPath(TestDescriptor descriptor) {
         return Joiner.on(" > ").join(displayedNames) + " ";
     }
 
-    private boolean isAtomicTestWhoseParentIsNotTheTestClass(TestDescriptor current) {
-        return !current.isComposite() && current.getClassName() != null && (current.getParent() == null
-                || !current.getClassName().equals(current.getParent().getName()));
+    private boolean isAtomicTestWhoseParentIsNotTheTestClassAndWhoseGrandparentHasNoTestClass(TestDescriptor current) {
```

Comment:
This method name seems... excessive. IMO the name should describe more what its trying to accomplish rather than what it's checking. For example, most of the text in this method name I would say belongs in javadoc, and the name itself would be more fitting as `shouldPrefixWithTaskName`.

The nature of this method is fishy. It seems to be trying to detect some really specific configuration of the test hierarchy. Is there somewhere upstream where can mark this case and detect that later here? I see the comment above mentions `testNG` is that still relevant? Does this circumstance only happen when we're running TestNG?

Reply:
Yeah, I agree the existing design isn't ideal. I think the TestDescriptor should be asked for its event path rather than the logger having these private methods which know implementation-specific details about the TestDescriptor (e.g. TestNG doesn't use class-level events).

I think the structural refactoring to improve that is out of scope for this bug fix. I'll add tests to make sure that there is no NPE, but will defer any design changes.